Add initial wiki linting

ddbeck opened the pull request:
This is the first step in linting and scraping the wiki. These changes were starting to get a bit big and I had to rework a bit when I realized I needed to consume the unrendered KumaScript macros, so I've decided to break the work of https://github.com/mdn/sprints/issues/2411 into pieces, to make the reviews more manageable. 

If you check out this PR, you can run `npm install` and then `node ./scripts/scraper-ng/ --help` to get started. If you provide an MDN page URL as an argument, you ought to get warnings for KumaScript macros and an explicit failure if you try to lint a page that doesn't have a compat macro, you'll get an error about that.

Things I haven't attempted to deal with yet:

* conversion to Markdown
* writing out Markdown
* making sense of page types and recipes

Feedback I'm especially interested in:

* can you imagine extending this? (e.g., writing more plugins to change the HTML tree or rules to lint the HTML content)? Can you imagine writing some checks for https://github.com/mdn/sprints/issues/2410?
* can you reason about the processing pipeline here? (e.g., is it clear how the script goes from from URL to inspecting the tree?)

But I'm curious what you think, in general, at this early stage.

Elchi3 commented on scripts/scraper-ng/index.js:
```diff
@@ -0,0 +1,58 @@
+const processor = rehype()
+  .use([require("./plugins/kumascript-macros")])
+  .use([
+    require("./rules/html-has-bcd-table"),
```
I think you renamed this at some point but not here.
```suggestion
    require("./rules/html-require-compat-macro"),
```

Elchi3 commented on scripts/scraper-ng/index.js:
```diff
@@ -0,0 +1,58 @@
+const toVFile = require("./url-to-vfile");
+
+const examplePage = "https://developer.mozilla.org/…/div";
+const examplePath = "https://developer.mozilla.org/…/div";
```
examplePath and examplePage are the same. Were you intending to support pages and page trees to scrape?
Also, this isn't a real URL, would it be better to use real examples?

ddbeck replied:
Yeah, an artifact of my development "process" 😆

I originally thought I'd allow you to supply a bunch of URLs at once. That's… complicated. I'm going to pare this back now, but we'll definitely need a bulk option. More on this in a bit.

ddbeck replied:
Oh wait, I misunderstood this. Originally, I was going to let you supply a full URL or a shorthand path (e.g., `https://developer.mozilla.org/en-US/docs/Web/HTML/Element/div` or `/en-US/docs/Web/HTML/Element/div`). Not doing that at all.

ddbeck replied:
OK, I've fixed the help text and the code to reflect what's actually happening with one URL at a time. I'll address the page trees question in a comment later.

Elchi3 commented on scripts/scraper-ng/rules/html-warn-macros.js:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Issue a warning for each macro
+ */
+function warnOnMacros(tree, file) {
```
This warns for the compat macro which we require in the other rule. Should we have a list of required / allow listed macros somewhere that both rules could use?

ddbeck replied:
Done. Also a nice demonstration of adding preferences to rules. 👍

ddbeck commented:
I've pushed a few changes addressing your comments, @Elchi3. Thank you!

> The lint-only mode is nice as well, but I would imagine that you want to support page trees as well. A fairly common use case will be to scrape/lint all pages under a given section (say all pages under ttps://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt).

Yes, this is a good point. One thing I haven't coped with yet is how to tell the linter what the rules should be for a given page. It'd be easy enough to map recipes to unified presets (e.g., recipe X has some list of applicable linting rules and other plugins), but I'm not sure how the linter should decide what page type it's looking at. A command line option is easy for one page but, as you say, you'll want to lint big chunks of content.

I've been thinking about a few strategies, but I'd be interested in hearing your thoughts on how you would expect the linter to behave.

wbamberg commented:
>  I'm not sure how the linter should decide what page type it's looking at

Yes, I had thought a little about that. The only thing I can think is, we will need to use tags, and make sure they are properly set for pages where that matters (that is, pages where we want to apply a more specific recipe than "this is a guide page").

ddbeck commented on scripts/scraper-ng/index.js:
```diff
@@ -0,0 +1,54 @@
+  .alias("q", "quiet")
+  .describe("v", "Expand linter notes, if applicable")
+  .alias("v", "verbose")
+  .boolean(["n", "q", "v"]).argv;
```
Yeah, that's about right. `--quiet` actually works if there's no warnings (e.g., if you comment out the macro warning), but the dry run and verbose options don't have a real effect yet. (The verbose option did, when I was scraping the fully rendered HTML and provided some help text for BCD tables without the right class name.)

ddbeck commented on scripts/scraper-ng/index.js:
```diff
@@ -0,0 +1,54 @@
+  .use([
+    require("./rules/html-require-compat-macro"),
+    [require("./rules/html-warn-macros"), ["Compat"]]
+  ]);
```
Yeah, I can change the way this is done a little bit. I imagine we'll use unified's [preset](https://github.com/unifiedjs/unified#preset) functionality to have different lists of plugins and configuration for different page types. I'll update the current PR to use a preset module though, which should help with readability.

wbamberg replied:
This is much nicer :)

ddbeck commented on scripts/scraper-ng/plugins/kumascript-macros.js:
```diff
@@ -0,0 +1,29 @@
+/*
+ * Find KumaScript macros in text literal nodes and put them in separate nodes
+ * with data about the macro calls.
+ */
```
Can you give me an example of only matching the first in each node? My initial experimentation with `hast-util-find-and-replace` had me thinking it was finding and replacing all instances within a text node. If it's not matching all of them, that's a bug and I should fix it.

ddbeck replied:
Nevermind, I found an example. I'll push a fix for this.

ddbeck commented on scripts/scraper-ng/plugins/kumascript-macros.js:
```diff
@@ -0,0 +1,29 @@
+ * with data about the macro calls.
+ */
+
+module.exports = kumascriptMacrosPlugin;
```
I'll fix this. I was cribbing from the unified examples and didn't think about this.

ddbeck commented on scripts/scraper-ng/plugins/kumascript-macros.js:
```diff
@@ -0,0 +1,29 @@
+    findAndReplace(tree, macroRegex, match => {
+      const groups = match.match(macroRegex);
+      const macroName = groups[1];
+      const macroParams = groups[2];
```
👍 Got it. I'll take care of this now.

ddbeck commented on scripts/scraper-ng/rules/html-warn-macros.js:
```diff
@@ -0,0 +1,26 @@
+const rule = require("unified-lint-rule");
+const visit = require("unist-util-visit-parents");
+
+module.exports = rule("html-warn-on-macros", warnOnMacros);
```
I didn't get as far with this as I would've liked, but my thinking was that some things might be easier to lint in other forms (e.g., you could lint rehype, then retext, then convert to Markdown). Or we might want to share some plugins between the scraper and the stumptown linter and it'd be easier if we identified what kind of tree the plugin works on.

ddbeck commented on scripts/scraper-ng/url-to-vfile.js:
```diff
@@ -0,0 +1,47 @@
+      f.contents = await response.text();
+    } else {
+      const message = f.message(`${response.status} ${response.statusText}`);
+      message.fatal = true;
```
Yeah, you're right there's a more tidy way to handle this. I'll fix this to call `f.fail` here (which throws).

ddbeck commented on scripts/scraper-ng/url-to-vfile.js:
```diff
@@ -0,0 +1,47 @@
+    }
+  } catch (err) {
+    f.message(err.message);
+    f.fatal = true;
```
Good catch. I'll apply this change.

ddbeck commented on scripts/scraper-ng/url-to-vfile.js:
```diff
@@ -0,0 +1,47 @@
+  return f;
+}
+
+function toRawMacros(url) {
```
I forgot to rename it when I changed what it did. 😳 I'll fix this.

ddbeck commented on scripts/scraper-ng/rules/html-require-compat-macro.js:
```diff
@@ -0,0 +1,25 @@
+const rule = require("unified-lint-rule");
+const visit = require("unist-util-visit");
+
+module.exports = rule("html-require-compat-macro", requireCompatMacro);
```
I wanted to demonstrate a simple macro that worked on the tree after I modified it. A real check on compat macros is, as you explained, more complicated. For guide pages we'll allow any number of compat macros and we'll rewrite them as stumptown directives. For JS method pages, for example, we'll want there to be one and only one compat macro under the "Browser compatibility" heading; we'll have specific scraping for that too (into frontmatter and to delete the body text).

But you're right that this check, as written, is general. I'll update it.

ddbeck commented:
Pushed a bunch of changes based on Will's comments (thank you, Will). Some notes:

- Moved linting rules config into `preset.js` module (I expect we'll have different presets for different page types, when we get to that level of sophistication)
- `preset.js` takes advantage of `unified-lint-rule`'s severity stuff. Slightly complicates the config, but breaking out into a separate module allows for variable names to spell things out better. I think this is an overall improvement, but I'd like feedback on that especially
- Fixed some macros not being found by the `kumascript-macros` plugin
- Parsed macro parameters
- Started printing macro call parameters for the html-warn-macros rule when the `--verbose` flag is used
- Better error handling on failed page loads (it's still complicated, since I have to handle non-200-300 responses and network errors separately, but it's better)
- Generalized required compat rule to be able to require any macro call
- Moved exports to the end of each file
- Expanded comment on `html-warn-macros`
- Renamed `toRawMacros` to `toRaw`

wbamberg commented on scripts/scraper-ng/url-to-vfile.js:
```diff
@@ -0,0 +1,49 @@
+      f.fail(`${response.status} ${response.statusText}`);
+    }
+  } catch (err) {
+    if (!isVFileMessage(err)) {
```
This check seems a bit non-obvious. My first thought was, "in what conditions will `try` throw without there being a fatal error"? But I think what's actually happening is that `fail()` is setting `fatal`. So really the error-handling code is still duplicated, except that one of the copies is hidden inside `fail()`. It would seem clearer to just throw in this case too, then there's only one place it gets handled. But it's a super-minor point, and probaby subjective anyway.

ddbeck replied:
I wasn't quite sure what you were getting at here but I think I rubber ducked my way to it (I've left my notes about this below, to show what I was thinking). I guess what should happen here is that I have a wrapper for `fetch` that either re-throws `fetch`'s exceptions or creates a new exception and then have the `vfile` error handling happen in only one place.

I'll push some changes for this shortly.

---

I agree it's a bit hard to follow but I'm not quite sure how to make it more obvious. Here's what I'm trying to capture with the existing code:

* The successful path: `fetch` succeeds at the network level and the response is successful (i.e., the page loads).
* An error response: `fetch` succeeds at the network level but the response is an error (e.g., the server responds with a 404). This corresponds to the `try` block's `else` condition. In the existing code, `f.fail()` throws and is swallowed by the `catch` block.
* Fetch failure: `fetch` itself throws (i.e., it can't make a connection to the server). In the existing code, a new error is created.

In the error cases, I want the `vfile` to have one and only one fatal error attached to it and then return the vfile (i.e., I never want to actually throw from `toVFile` at all).

ddbeck commented:
@wbamberg Thank you! I think I finally figured out what you meant by the fetch handling. Could you take another look at that?

wbamberg commented:
Yes, that's exactly what I meant :). Thanks Daniel!

ddbeck commented:
@Elchi3 Did you want another look at this before merging?

ddbeck commented:
Thank you both, @Elchi3 and @wbamberg. I think this is a lot better than it was originally. I'll get cracking on the output side of things soon.
